### PR TITLE
vulkan: detect multiple devices by deviceUUID instead of deviceID

### PR DESCRIPTION
**Pull request by Adriankhl**
Fixes https://github.com/ggerganov/llama.cpp/issues/7997

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [X] Low
  - [ ] Medium
  - [ ] High

### CODE REVIEW

File: ggml-vulkan.cpp [hunk 1]
```diff
@@ -1745,39 +1745,45 @@ void ggml_vk_instance_init() {
 
         // Default to using all dedicated GPUs
         for (size_t i = 0; i < devices.size(); i++) {
-            vk::PhysicalDeviceProperties props = devices[i].getProperties();
-
-            if (props.deviceType == vk::PhysicalDeviceType::eDiscreteGpu) {
+            vk::PhysicalDeviceProperties new_props_1 = devices[i].getProperties();
```
Comment (0cc4m) [inline ggml-vulkan.cpp:1748]: This isn't needed, you can get the PhysicalDeviceProperties from `new_props_2.properties`. Then you also don't need to call it "new_props_2`, just `new_props`

File: ggml-vulkan.cpp [hunk 2]
```diff
@@ -1745,39 +1745,45 @@ void ggml_vk_instance_init() {
...
+            devices[i].getProperties2(&new_props_2);
+
+            if (new_props_1.deviceType == vk::PhysicalDeviceType::eDiscreteGpu) {
                 // Check if there are two physical devices corresponding to the same GPU
                 auto old_device = std::find_if(
                     vk_instance.device_indices.begin(),
                     vk_instance.device_indices.end(),
-                    [&devices, &props](const size_t k){ return devices[k].getProperties().deviceID == props.deviceID; }
+                    [&devices, &new_id](const size_t k){ 
+                        vk::PhysicalDeviceProperties2 old_props_2;
```
Comment (0cc4m) [inline ggml-vulkan.cpp:1762]: No need for the 2, `old_props`

File: ggml-vulkan.cpp [hunk 3]
```diff
@@ -1745,39 +1745,45 @@ void ggml_vk_instance_init() {
...
                 if (old_device == vk_instance.device_indices.end()) {
                     vk_instance.device_indices.push_back(i);
                 } else {
                     // There can be two physical devices corresponding to the same GPU if there are 2 different drivers
                     // This can cause error when splitting layers aross the devices, need to keep only 1
-                    VK_LOG_DEBUG("Device " << i << " and device " << *old_device << " have the same device id");
+                    VK_LOG_DEBUG("Device " << i << " and device " << *old_device << " have the same deviceUUID");
 
-                    vk::PhysicalDeviceProperties2 old_prop;
+                    vk::PhysicalDeviceProperties2 old_props_2;
```
Comment (0cc4m) [inline ggml-vulkan.cpp:1775]: See above

**Review by 0cc4m — CHANGES_REQUESTED** [review]
I tested it and it works. I just have a few small comments about the code style, once you addressed those this can be merged.

**Comment by Adriankhl** [comment]
@0cc4m ok done

File: ggml-vulkan.cpp [hunk 4]
```diff
@@ -1745,39 +1745,44 @@ void ggml_vk_instance_init() {
...
+
+            if (new_props.properties.deviceType == vk::PhysicalDeviceType::eDiscreteGpu) {
                 // Check if there are two physical devices corresponding to the same GPU
                 auto old_device = std::find_if(
                     vk_instance.device_indices.begin(),
                     vk_instance.device_indices.end(),
-                    [&devices, &props](const size_t k){ return devices[k].getProperties().deviceID == props.deviceID; }
+                    [&devices, &new_id](const size_t k){
+                        vk::PhysicalDeviceProperties2 old_props;
+                        vk::PhysicalDeviceIDProperties old_id;
```
Comment (0cc4m) [inline ggml-vulkan.cpp:1762]:
```suggestion
                        vk::PhysicalDeviceIDProperties old_id;
                        old_props.pNext = &old_id;
```

**Review by 0cc4m — CHANGES_REQUESTED** [review]
I'm not sure how I missed that, it seemed to be working before. You need to actually fill the old_id.

**Review by 0cc4m — APPROVED** [review]
Thank you, now it's working as intended!

**Pull request merged**